whisper.cpp

```diff
@@ -2928,6 +2929,191 @@ static bool log_mel_spectrogram(
     return true;
 }
 
+// Algorithm for byte pair encoding
+// BPE essentially first breaks down a string into bytes
+// and then merges adjacent bytes together according to rules
+// until they can no longer be merged
+// This algo is for educational purposes and not optimized for high performance
+static std::vector<whisper_vocab::id> bpe_encode(const whisper_vocab & vocab, const std::string & word) {
+    std::vector<whisper_vocab::id> tokens;
+    tokens.reserve(word.size());
+
+    // split each word into an array of single byte
+    for (int pos=0; pos < word.size(); pos++) {
+        tokens.push_back(vocab.token_to_id.at(word.substr(pos, 1)));
+    }
+
+    while (true) {
+        int min_idx = -1;
+        int min_rank = -1;
+
+
+        // iterate over all pairs and find the pair we want to merge the most
+        for (int pos=0; pos < tokens.size() - 1; pos++) {
+            auto query = vocab.id_to_token.at(tokens[pos]) + vocab.id_to_token.at(tokens[pos+1]);
+            auto it = vocab.token_to_id.find(query);
+            if (it != vocab.token_to_id.end()) {
+                auto rank = it->second;
+                // find the pair with the lowest rank
+                if (min_rank == -1 || rank < min_rank) {
+                    min_idx = pos;
+                    min_rank = rank;
+                }
+            }
+        }
+
+        // if there were no pairs we could merge, we're done!
+        if (min_rank == -1) {
+            break;
+        }
+
+        // update token vector
+        tokens[min_idx] = min_rank;
+        tokens.erase(tokens.begin() + min_idx + 1);
+    }
+
+    return tokens;
+}
+
+// This is not perfect
+// Occasionally, it produces different results compared to OpenAI's tiktoken
```

Comment:
Can you give an example where the results don't match?

Reply:
When I directly used the implementation of `bpe_gpt2_preprocess` in llama.cpp, there was a `0.06%` error, which disappeared after modification. I should delete this comment.